### PR TITLE
feat(grid): add ux grid component

### LINKED ISSUE #138: ux-grid
Add UX Grid component for Aurelia UX and utility classes for responsive applications. Based on a 12 column system.

**Breakpoints**
Extra Small - 480px and smaller.
Small - Larger than 480px, smaller than 960px.
Medium - Larger than 960px, smaller than 1280px.
Large - Larger than 1280px, smaller than 1925px.
Extra Large - Larger than 1925px.

``` html
<ux-grid>
    <ux-grid-cell sm="12" md="6" lg="4">
        Side content
    </ux-grid-cell>
    <ux-grid-cell sm="12" md="6" lg="8">
        Main content
    </ux-grid-cell>
</ux-grid>
```

[Material Guidelines - Responsive UI](https://material.io/guidelines/layout/responsive-ui.html)

### PR DESCRIPTION
**Pull request by ZHollingshead**
Adds the ux-grid component and utilitiy functions.

**Breakpoints**
Extra Small - 480px and smaller.
Small - Larger than 480px, smaller than 960px.
Medium - Larger than 960px, smaller than 1280px.
Large - Larger than 1280px, smaller than 1925px.
Extra Large - Larger than 1925px.

**Usage**

*grid*
``` html
<ux-grid>
    <ux-grid-cell sm="12" md="6" lg="4">
        Side content
    </ux-grid-cell>
    <ux-grid-cell sm="12" md="6" lg="8">
        Main content
    </ux-grid-cell>
</ux-grid>
```

Full width grid
``` html
<!-- alternatively you can use class="fluid" -->
<ux-grid fluid>
    <ux-grid-cell sm="12">
        Full width grid
    </ux-grid-cell>
</ux-grid>
```

Remove padding from grid, does not remove gutters
``` html
<!-- alternatively you can use class="remove-padding" -->
<ux-grid remove-padding>
    <ux-grid-cell sm="12">
        Full width grid
    </ux-grid-cell>
</ux-grid>
```

*responsive utilities*
```ts
import { autoinject } from 'aurelia-dependency-injection';
import { UxResponsiveUtilities } from '@aurelia-ux/grid';

@autoinject
export class MyComponent {
  constructor(public responsive: ResponsiveUtilities) {}
}
```

```html
<div if.bind="responsive.xs || responsive.sm">
  small screens only
</div>

<div if.bind="responsive.md">
  medium screens only
</div>

<div if.bind="!responsive.xl">
  not on extra large screens
</div>
```

**Other Grid Classes**
`ux-grid.align-cells-top`, `ux-grid.align-cells-middle`, and `ux-grid.align-cells-bottom` can be used to align the cells in their respective rows relative to the other cells on that row. By default the grid stretches the cells.

`align-self-start`, `align-self-middle`, `align-self-bottom`, and `align-self-stretch` can be used to do the same as above, but to a single cell instead. This is applied directly to the `ux-grid-cell` tag.

`order`, `order-xs`, `order-sm`, `order-md`, `order-lg`, and `order-xl` can be used to change the order of the cell in the grid.

**Images**
image

image

resolves #138

### CODE REVIEW

**Comment by bigopon** [comment]
Looks very nice 👍 

For the `ux.sm`, `ux.lg` I think we know exactly when it's changed so we can extract all logic into a resize handler:

```js
onResize() {
  let width = getWidth();
  let height = getHeight();

  this.sm = width < 480;
  this.md = width > 480 * width < 640;
  // etc ...
}
```

**Comment by ZHollingshead** [comment]
Thanks for the advice, I made that change and implemented disposable on the `ResponsiveUtilities` class.

File: packages/grid/src/ux-responsive-utilities.ts [hunk 1]
```diff
@@ -1,48 +1,58 @@
...
-    return this.width <= 480;
-  }
+  private recentlyUpdated = false;
 
-  /** Visible on screens larger than 480px, and smaller than 960px. */
-  public get sm() {
-    return this.width > 480 && this.width <= 960;
-  }
+  constructor() {
+    window.addEventListener('resize', () => this.calculateResponsiveValues());
```
Comment (davismj) [inline packages/grid/src/ux-responsive-utilities.ts]: This needs to use RAF or at least some kind of debouncing.

File: packages/grid/src/ux-responsive-utilities.ts [hunk 2]
```diff
@@ -1,48 +1,58 @@
...
+    this.height = PLATFORM.global.innerHeight;
+    this.width = PLATFORM.global.innerWidth;
+
+    this.xs = this.width <= 480;
+    this.sm = this.width > 480 && this.width <= 960;
+    this.md = this.width > 960 && this.width <= 1280;
+    this.lg = this.width > 1280 && this.width <= 1925;
+    this.xl = this.width > 1925;
+
+    setTimeout(() => this.recentlyUpdated = false, 250);
```
Comment (davismj) [inline packages/grid/src/ux-responsive-utilities.ts]: Why not polyfill RAF instead? Why 250 throttle instead of a 100 second debounce?

**Comment by ZHollingshead** [comment]
Are there any other suggestions?

I will have full documentation for this component and a usage example with the new app-ux-showcase update I am working on which has had its layout completely revamped to support this.

@EisenbergEffect what are your thoughts?

**Comment by EisenbergEffect** [comment]
@ZHollingshead Is the entire layout of the new demo app built with this grid system?

**Comment by ZHollingshead** [comment]
@EisenbergEffect No, the Toolbar/Header and Sidebar are static pieces still, but everything within the Main element will use the grid as its layout.

**Comment by ZHollingshead** [comment]
@EisenbergEffect
app.html
```html
<template>
  <require from="./app.css"></require>

  <ux-drawer type="permanent"></ux-drawer>

  <main>
    <ux-toolbar></ux-toolbar>

    <router-view class="page"></router-view>
  </main>
</template>
```

forms.html
```html
<template>
  <require from="./forms.css"></require>

  <ux-grid fluid>
    <ux-grid-cell>
      <h1>
        &lt;ux-form&gt;&lt;/ux-form&gt;
      </h1>

      <p>
        The <code>ux-form</code> element is used to quickly create and lay out forms.
      </p>
    </ux-grid-cell>

    <ux-grid-cell sm="6" md="4">
      <!-- example code -->
    </ux-grid-cell>
    <ux-grid-cell sm="6" md="4">
      <!-- example code -->
    </ux-grid-cell>


    <ux-grid-cell>
      <h3>Form Row</h3>

      <p>
        The <code>ux-form</code> includes the <code>.form-row</code> class to help with laying out a form and including multiple controls on one line.
      </p>
    </ux-grid-cell>

    <ux-grid-cell sm="6" md="4">
      <!-- example code -->
    </ux-grid-cell>
  </ux-grid>
</template>
```

**Comment by EisenbergEffect** [comment]
Cool.

**Comment by ZHollingshead** [comment]
@EisenbergEffect Do you see anything else or is this good to go?

**Comment by EisenbergEffect** [comment]
Let's go with it and see how it pans out in the new demo app and in combination with other components. I'd love to see how the community uses it.

**Pull request merged**